webapp/src/ts/services/training-cards.service.ts

```diff
@@ -0,0 +1,124 @@
+import { Injectable } from '@angular/core';
+import { Store } from '@ngrx/store';
+
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { TrainingCardsComponent } from '@mm-modals/training-cards/training-cards.component';
+import { DbService } from '@mm-services/db.service';
+import { GlobalActions } from '@mm-actions/global';
+import { ModalService } from '@mm-modals/mm-modal/mm-modal';
+import { SessionService } from '@mm-services/session.service';
+import { RouteSnapshotService } from '@mm-services/route-snapshot.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TrainingCardsService {
+  private globalActions;
+  private readonly TRAINING_PREFIX = 'training:'
+
+  constructor(
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private dbService: DbService,
+    private modalService: ModalService,
+    private sessionService: SessionService,
+    private routeSnapshotService: RouteSnapshotService,
+  ) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  private canOpenTraining() {
+    const routeSnapshot = this.routeSnapshotService.get();
+    return !this.sessionService.isDbAdmin() &&
+      (!document.getElementsByClassName('enketo').length ||
+        (routeSnapshot?.data?.tab === 'tasks' && !routeSnapshot?.params?.id));
+  }
+
+  private getAvailableTrainingForms(xForms, userCtx) {
+    const today = new Date();
+
+    return xForms
+      .map(xForm => ({
+        id: xForm._id,
+        code: xForm.internalId,
+        startDate: xForm.start_date,
+        duration: xForm.duration,
+        userRoles: xForm.user_roles,
+      }))
+      .filter(form => {
+        const hasRole = form.userRoles?.find(role => userCtx.roles.includes(role));
+        if (form.userRoles?.length && !hasRole) {
```

Comment:
You don't need to check the length - if it's undefined or empty then `hasRole` will be falsey, right?

```suggestion
        if (!hasRole) {
```